include/aspect/simulator/assemblers/interface.h

```diff
@@ -0,0 +1,718 @@
+/*
+  Copyright (C) 2017 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef _aspect_simulator_assemblers_interface_h
+#define _aspect_simulator_assemblers_interface_h
+
+#include <aspect/global.h>
+#include <aspect/heating_model/interface.h>
+#include <aspect/material_model/interface.h>
+
+#include <deal.II/fe/fe_values.h>
+
+namespace aspect
+{
+  using namespace dealii;
+
+  template <int dim>
+  class Simulator;
+
+  struct AdvectionField;
+
+  /**
+   * A namespace that is used for internal scratch objects, i.e. objects that define
+   * the inputs and outputs for the individual assembler objects. The different classes
+   * are provided for the different systems of equations, and the inputs are filled
+   * by the <code>local_assemble_...</code> functions in <code>assembly.cc</code>,
+   * before being handed over to the assemblers.
+   */
+  namespace internal
+  {
+    namespace Assembly
+    {
+      namespace Scratch
```

Comment:
My vote would be to remove this namespace as it doesn't convey additional information (``Scratch::ScratchBase`` vs ``ScratchBase``)